gtsam/inference/EliminateableFactorGraph-inst.h

```diff
@@ -301,29 +341,53 @@ namespace gtsam {
     } else {
       // No ordering was provided for the marginalized variables, so order them using constrained
       // COLAMD.
-      bool unmarginalizedAreOrdered = (boost::get<const Ordering&>(&variables) != 0);
-      const KeyVector* variablesOrOrdering =
-        unmarginalizedAreOrdered ?
-        boost::get<const Ordering&>(&variables) : boost::get<const KeyVector&>(&variables);
+      constexpr bool forceOrder = true;
+      Ordering totalOrdering =
+        Ordering::ColamdConstrainedLast((*variableIndex).get(), variables, forceOrder);
+
+      // Split up ordering
+      const size_t nVars = variables.size();
+      Ordering marginalizationOrdering(totalOrdering.begin(), totalOrdering.end() - nVars);
+      Ordering marginalVarsOrdering(totalOrdering.end() - nVars, totalOrdering.end());
+
+      // Call this function again with the computed orderings
+      return marginalMultifrontalBayesTree(std::cref(marginalVarsOrdering), marginalizationOrdering, function, variableIndex);
+    }
+  }
 
+  /* ************************************************************************* */
+  template<class FACTORGRAPH>
+  std::shared_ptr<typename EliminateableFactorGraph<FACTORGRAPH>::BayesTreeType>
+    EliminateableFactorGraph<FACTORGRAPH>::marginalMultifrontalBayesTree(
+    const KeyVector& variables,
+    const Eliminate& function, OptionalVariableIndex variableIndex) const
+  {
+    if(!variableIndex) {
+      // If no variable index is provided, compute one and call this function again
+      VariableIndex computedVariableIndex(asDerived());
+      return marginalMultifrontalBayesTree(variables, function, std::cref(computedVariableIndex));
+    } else {
+      // No ordering was provided for the marginalized variables, so order them using constrained
+      // COLAMD.
+      constexpr bool forceOrder = false;
       Ordering totalOrdering =
-        Ordering::ColamdConstrainedLast((*variableIndex).get(), *variablesOrOrdering, unmarginalizedAreOrdered);
+        Ordering::ColamdConstrainedLast((*variableIndex).get(), variables, forceOrder);
 
       // Split up ordering
-      const size_t nVars = variablesOrOrdering->size();
+      const size_t nVars = variables.size();
       Ordering marginalizationOrdering(totalOrdering.begin(), totalOrdering.end() - nVars);
       Ordering marginalVarsOrdering(totalOrdering.end() - nVars, totalOrdering.end());
 
       // Call this function again with the computed orderings
-      return marginalMultifrontalBayesTree(marginalVarsOrdering, marginalizationOrdering, function, variableIndex);
+      return marginalMultifrontalBayesTree(std::cref(marginalVarsOrdering), marginalizationOrdering, function, variableIndex);
```

Comment:
We can remove `cref` since we don't use reference_wrapper anymore

Reply:
Good catch !!!